Luca joss/memo and receiver filtering

ljoss17 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #XXX

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->
<!-- Apply relevant labels to indicate:
    - (WHY) The purpose or objective of this PR with "O" labels
    - (WHICH) The part of the system this PR relates to (use "E" for external or "I" for internal levels)
    - (HOW) If any administrative considerations should be taken into account (use "A" labels)
    This will help us prioritize and categorize your pull request more effectively 
-->


______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

romac commented on crates/relayer/src/config/types.rs:
```diff
@@ -175,6 +179,164 @@ pub mod max_tx_size {
+        where
+            D: Deserializer<'de>,
+        {
+            let value = usize::deserialize(deserializer)?;
```
Let's use `byte_unit::Byte` like we do for `max_grpc_decoding_size`.

ljoss17 replied:
Fixed in 43a79a35c048b1eeef4d50114b2f2e060154fb4e

romac commented on crates/relayer/src/config/types.rs:
```diff
@@ -175,6 +179,164 @@ pub mod max_tx_size {
+        where
+            D: Deserializer<'de>,
+        {
+            let value = usize::deserialize(deserializer)?;
```
Let's use `byte_unit::Byte` like we do for `max_grpc_decoding_size`.

ljoss17 replied:
Fixed in 43a79a35c048b1eeef4d50114b2f2e060154fb4e

romac commented on crates/relayer/src/link.rs:
```diff
@@ -43,9 +45,16 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> Link<ChainA, ChainB> {
     pub fn new(
         channel: Channel<ChainA, ChainB>,
         with_tx_confirmation: bool,
+        max_memo_size: usize,
```
Let's maybe pass and store the whole `LinkParameters` struct here instead of each field? What do you think?

ljoss17 replied:
Good point, it is cleaner this way

ljoss17 replied:
Fixed in f2b3a4fc46891f7f047dfb3a6336bf2a31409b59

romac commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -108,12 +108,17 @@ pub struct RelayPath<ChainA: ChainHandle, ChainB: ChainHandle> {
     pub fn new(
         channel: Channel<ChainA, ChainB>,
         with_tx_confirmation: bool,
+        max_memo_size: usize,
```
Same here

ljoss17 replied:
Fixed in f2b3a4fc46891f7f047dfb3a6336bf2a31409b59

romac commented on crates/relayer-types/src/core/ics04_channel/packet.rs:
```diff
@@ -178,6 +181,32 @@ impl Packet {
         height_timed_out || timestamp_timed_out
     }
+
+    pub fn validate_fields(
```
Let's add some doc comments to explain what it does, with a couple examples perhaps?

romac replied:
Also, the semantics of this function are a bit weird.

Instead of returning the size when it's too big and 0 when it's okay, how about returning a bespoke `Error` type like:

```rust
enum ValidationError {
  MemoTooBig { size: usize, max: usize },
  ReceiverTooBig { size: usize, max: usize },
  Other(Error),
}
```

and returning the actual sizes when all is okay?

romac commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -1381,11 +1389,31 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+                    Ok((None, None))
+                }
+                Err(e) => {
+                    warn!("error validating fields, will simply relay message: {e}");
+                    Ok((self.build_recv_packet(&event.packet, height)?, None))
```
Will this warning show for all non-ICS-20 packets? That would be a bit noisy.

We should only show a warning/error if we see that we have an ICS-20 packet and:
- are unable to decode it
- the memo or the receiver fields are too large

What do you think?

romac commented on crates/relayer-types/src/core/ics04_channel/packet.rs:
```diff
@@ -178,6 +181,32 @@ impl Packet {
+    ) -> Result<(usize, usize), Error> {
+        let any_packet: RawPacketData =
+            serde_json::from_slice(&self.data).map_err(Error::serde_json_error)?;
+        let packet: Ics20PacketData = any_packet
+            .try_into()
+            .map_err(|e: Ics20Error| Error::decode_ics20_packet(e.to_string()))?;
```
This is probably overkill, what do you think about just looking at the fields of the RawPacketData instead?

ljoss17 replied:
Good point I will remove this

romac commented on crates/relayer-types/src/core/ics04_channel/packet.rs:
```diff
@@ -178,6 +181,32 @@ impl Packet {
+        max_receiver_size: usize,
+    ) -> Result<(usize, usize), Error> {
+        let any_packet: RawPacketData =
+            serde_json::from_slice(&self.data).map_err(Error::serde_json_error)?;
```
Here we should probably not return an error if the packet data is not an ICS-20 packet, but rather return `Ok((0, 0))` or something like that.

romac commented on crates/relayer-types/src/core/ics04_channel/packet.rs:
```diff
@@ -178,6 +178,19 @@ impl Packet {
         height_timed_out || timestamp_timed_out
     }
+
+    /// This method will return if the memo and receiver fields are valid, e.g. not too big.
+    /// If it fails to decode the ICS20 PacketData then there is no need to validate fields
+    /// so the method will return true.
+    pub fn are_fields_valid(&self, max_memo_size: u64, max_receiver_size: u64) -> bool {
```
```suggestion
    /// If this packet data is an ICS-20 packet, check whether or not the memo and receiver fields are under the given max size.
    /// Returns `true` if this packet data is not an ICS-20 or fails to decode as an ICS-20 packet.
    pub fn check_ics20_fields_size(&self, max_memo_size: usize, max_receiver_size: usize) -> bool {
```

romac commented on crates/relayer-cli/src/commands/clear.rs:
```diff
@@ -135,6 +135,8 @@ impl Runnable for ClearPacketsCmd {
         let opts = LinkParameters {
             src_port_id: self.port_id.clone(),
             src_channel_id: self.channel_id.clone(),
+            max_memo_size: config.mode.packets.ics20_max_memo_size.get_bytes(),
+            max_receiver_size: config.mode.packets.ics20_max_receiver_size.get_bytes(),
```
```suggestion
            max_memo_size: config.mode.packets.ics20_max_memo_size.get_bytes() as usize,
            max_receiver_size: config.mode.packets.ics20_max_receiver_size.get_bytes() as usize,
```

ljoss17 replied:
Fixed in 0fd448cd8e91d9a825357093c9694cc2e817e6af

romac commented on crates/relayer-types/src/core/ics04_channel/packet.rs:
```diff
@@ -178,6 +178,19 @@ impl Packet {
+        match serde_json::from_slice::<RawPacketData>(&self.data).map_err(Error::serde_json_error) {
+            Ok(packet) => {
+                packet.memo.len() <= max_memo_size as usize
+                    || packet.receiver.to_string().len() <= max_receiver_size as usize
```
Shouldn't this be a conjunction?

```suggestion
                    && packet.receiver.to_string().len() <= max_receiver_size as usize
```

romac replied:
Also, let's avoid an allocation here:

```suggestion
                    && packet.receiver.as_ref().len() <= max_receiver_size as usize
```

ljoss17 replied:
Oh right, since I inverted the assertion it needs to be `&&`, good catch!

romac commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -1381,11 +1389,27 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+        {
             Ok((self.build_recv_packet(&event.packet, height)?, None))
+        } else {
+            debug!(
+                "memo and/or receiver field are too big for packet {:#?}",
+                event.packet
+            );
```
Rather than outputting the whole packet here (which may be extremely large and flood the console), let's instead log the memo and receiver sizes and the max allowed sizes.

For this you might have to change the validation function again to return both sizes.

romac replied:
We can perhaps log the whole packet with `trace!`. What do you think @ljoss17  @ancazamfir?

ljoss17 replied:
I wanted to keep the validation method as simple as possible, and since it's a debug log I would be inclined to have more details in the logs. I could also put it as trace.

Or would you still prefer to have only the field sizes in the log outputs?

romac replied:
it would be nice to have the sequence number and field sizes in the debug log when we choose to not relay the packet. Will surely help debugging along the line.

Let's dump the whole packet data in `trace!` mode.

ljoss17 replied:
I introduced more advanced logs in this commit 9af4e6d0688748f09be91d09b22c7f272e82fb8c.
With debug logs when one or both fields fails and trace logs of the packet when there is any failure.

romac commented on config.toml:
```diff
@@ -79,6 +79,18 @@ tx_confirmation = false
+# Set the maximum size for the memo field in ICS20 packets.
+# If the size of the memo field is bigger than the configured
+# one, the packet will not be relayed.
+# [Default: 32768 (32Kb)]
+#ics20_max_memo_size = 32768
```
```suggestion
# [Default: "32KiB"]
#ics20_max_memo_size = "32KiB"
```

romac commented on config.toml:
```diff
@@ -79,6 +79,18 @@ tx_confirmation = false
+# Set the maximum size for the receiver field in ICS20 packets.
+# If the size of the receiver field is bigger than the configured
+# one, the packet will not be relayed.
+# [Default: 32768 (2Kb)]
+#ics20_max_receiver_size = 2048
```
```suggestion
# [Default: "2KiB"]
#ics20_max_receiver_size = "2KiB"
```

ancazamfir commented on config.toml:
```diff
@@ -79,6 +79,18 @@ tx_confirmation = false
 # [Default: false]
 auto_register_counterparty_payee = false
 
+# Set the maximum size for the memo field in ICS20 packets.
+# If the size of the memo field is bigger than the configured
+# one, the packet will not be relayed.
+# [Default: "32KiB"]
+#ics20_max_memo_size = "32KiB"
+
+# Set the maximum size for the receiver field in ICS20 packets.
+# If the size of the receiver field is bigger than the configured
+# one, the packet will not be relayed.
+# [Default: "2KiB"]
+#ics20_max_receiver_size = "2KiB"
```
Can you add config on how to achieve old behaviour, i.e. no checks?

ljoss17 replied:
Good point, I'll slightly change the configuration to have something along the lines of:

`ics20_max_receiver_limit = { enabled = true, size = "2KiB" }`
`ics20_max_memo_limit = { enabled = true, size = "32KiB" }`

ancazamfir commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -1381,11 +1389,55 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
-            Ok((self.build_recv_packet(&event.packet, height)?, None))
+            match event
+                .packet
+                .are_fields_valid(self.max_memo_size, self.max_receiver_size)
```
Why are we not checking this earlier, i.e. for timeouts also? Same question for `build_ack_from_recv_event()`? Maybe even in the supervisor?

ancazamfir replied:
Thinking about it more, while it would be nice to have this in the supervisor it would not cover the startup and periodic clearing. So I guess we can check in `relay_path` but let's decide if we want to also apply for the other (than recv) message types.

romac replied:
The check is now performed directly in `generate_operational_data` for all events which pertain to an ICS-04 packet: https://github.com/informalsystems/hermes/pull/3765/commits/3d5687eb534dc72a94c08e95e7afdfb6492d3d22

ancazamfir commented:
Left a few comments. In general I am wondering why we don't apply the checks for all msgs that contain a packet, i.e. recv, timeout and ack.

romac commented:
Good catch, we should indeed do that! Completely slipped my mind…

romac commented on crates/relayer-types/src/core/ics04_channel/packet.rs:
```diff
@@ -178,6 +199,48 @@ impl Packet {
+    /// Returns ValidationResult::Valid if both the memo and receiver field are valid.
+    /// Returns ValidationResult containing information on the failed fields if one or both fields fail
+    /// the validation.
+    pub fn are_fields_valid(
```
Let's pull this out of the `Packet` struct as it is specific to ICS-20 and have it instead within the ics20 module as a free-standing function, perhaps named `check_ics20_fields_length`

romac commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -1381,11 +1389,55 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+                    Ok((self.build_recv_packet(&event.packet, height)?, None))
+                }
+                ValidationResult::MemoTooBig { size, max } => {
+                    trace!("packet: {:#?}", event.packet);
```
```suggestion
                    trace!("ICS-20 packet failed fields length check: {:#?}", event.packet);
```

romac commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -1381,11 +1389,55 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+                    Ok((None, None))
+                }
+                ValidationResult::ReceiverTooBig { size, max } => {
+                    trace!("packet: {:#?}", event.packet);
```
```suggestion
                    trace!("ICS-20 packet failed fields length check: {:#?}", event.packet);
```

romac commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -1381,11 +1389,55 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+                    receiver_size,
+                    receiver_max,
+                } => {
+                    trace!("packet: {:#?}", event.packet);
```
```suggestion
                    trace!("ICS-20 packet failed fields length check: {:#?}", event.packet);
```

romac commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -1381,11 +1389,55 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+                    Ok((None, None))
+                }
+                ValidationResult::DecodeFailure { details } => {
+                    trace!("packet: {:#?}", event.packet);
```
I think we can remove this in that case

romac commented on crates/relayer/src/link/relay_path.rs:
```diff
@@ -1381,11 +1389,55 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+                }
+                ValidationResult::DecodeFailure { details } => {
+                    trace!("packet: {:#?}", event.packet);
+                    trace!("failed to decode packet or packet isn't an ICS20 packet, details: {details}");
```
Is there a way to determine whether or not the packet was supposed be a valid ICS-20 packet?

ljoss17 commented:
I refactored the memo and receiver field configuration and validation in this commit 451b5781f809fd6bf13b1c66340a67a0e4bc6b27.

I added the option to disable the filter by having the configuration format as following: `{ enabled = true, size = "32KiB" }`

The reason why I put the `Ics20FieldSizeLimit` struct in the `relayer` crate is because I feel like validation linked to the configuration shouldn't be in the `relayer-types` crate.

I only applied the verification https://github.com/informalsystems/hermes/pull/3765/commits/451b5781f809fd6bf13b1c66340a67a0e4bc6b27#diff-072b872f1fbf4ea17384cecde28b35ee339faa9ce969ff9f767647a071da1e36R1405 as I wanted to have your opinion on this solution @ancazamfir @romac. If this seems like a good solution we can think more in detail to where we want to validate these fields

romac commented:
The check is now performed directly in `generate_operational_data` for all events which pertain to an ICS-04 packet: https://github.com/informalsystems/hermes/commit/3d5687eb534dc72a94c08e95e7afdfb6492d3d22

romac commented:
The guide template checker is failing but that's unrelated to this PR. Will fix out of band.

ancazamfir reviewed (approved):
Looks good! Thanks @ljoss17!
